### PR TITLE
[AST] Store SubstitutionMaps in the AST rather than Substitution lists.

### PR DESCRIPTION
**Pull request by DougGregor**
Consistently store `SubstitutionMap` within various parts of the AST rather than `Substitution` lists, and deal with the fallout. Specific changes:

* `ConcreteDeclRef` now uses `SubstitutionMap` and has been simplified.
* `Witness` now uses `SubstitutionMap`
* `SpecializedProtocolConformance` now uses `SubstitutionMap`
* `SILBoxType` now uses `SubstitutionMap`.

Overall, removes ~90 explicit uses of `SubstitutionList` (of ~410).

### CODE REVIEW

**Comment by slavapestov** [comment]
Yay!

File: include/swift/SIL/SILFunction.h [hunk 1]
```diff
@@ -123,6 +123,9 @@ class SILFunction
   /// The forwarding substitutions, lazily computed.
   Optional<SubstitutionList> ForwardingSubs;
 
+  /// The forwarding substitution map, lazily computed.
+  SubstitutionMap ForwardingSubMap;
```
Comment (slavapestov) [inline include/swift/SIL/SILFunction.h]: Perhaps we should move this caching logic to GenericEnvironment?

File: lib/AST/ASTVerifier.cpp [hunk 2]
```diff
@@ -2416,6 +2416,10 @@ class Verifier : public ASTWalker {
       }
     }
 
+    void verifyChecked(SubstitutionMap substitutions){
+      // FIXME: Check replacement types without forcing anything.
```
Comment (slavapestov) [inline lib/AST/ASTVerifier.cpp]: SubstitutionMap has a verify() method

File: lib/SILGen/SILGenApply.cpp [hunk 3]
```diff
@@ -4805,11 +4812,19 @@ emitSpecializedAccessorFunctionRef(SILGenFunction &SGF,
                                    bool isSuper,
                                    bool isDirectUse)
 {
+  // Convert the substitution list into a substitution map.
+  auto origAccessType = SGF.SGM.Types.getConstantInfo(constant).FormalType;
+  SubstitutionMap subMap;
+  if (auto genericFnType = dyn_cast<GenericFunctionType>(origAccessType)) {
+    auto genericSig = genericFnType.getGenericSignature();
+    subMap = genericSig->getSubstitutionMap(substitutions);;
```
Comment (slavapestov) [inline lib/SILGen/SILGenApply.cpp]: Unnecessary ;, and this could use toList() instead

File: lib/Sema/TypeCheckDecl.cpp [hunk 4]
```diff
@@ -3371,27 +3366,44 @@ static void checkVarBehavior(VarDecl *decl, TypeChecker &TC) {
...
-    Substitution(decl->getInterfaceType(), valueSub.getConformances()),
-  };
-  Substitution allContextSubs[] = {
-    Substitution(behaviorSelf, *selfConformance),
-    Substitution(decl->getType(), valueSub.getConformances()),
-  };
+  // Build interface and context substitution maps.
+  auto protoSelfType = behaviorProto->getProtocolSelfType();
+  auto sig = behaviorProto->getGenericSignatureOfContext();
+  auto interfaceSubsMap =
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: This is SubstitutionMap::getProtocolSubstitutions()

**Review by slavapestov — COMMENTED** [review]
I have a few suggestions after taking another look :-)

File: include/swift/SIL/SILFunction.h (hunk 1 above)
Comment (DougGregor) [reply include/swift/SIL/SILFunction.h]: So little code cares about it that it doesn't seem necessary.

File: lib/SILGen/SILGenApply.cpp (hunk 3 above)
Comment (DougGregor) [reply lib/SILGen/SILGenApply.cpp]: It's going the other way, `SubstitutionList` -> `SubstitutionMap`. I'll remove the extra `;` of course.

File: lib/Sema/TypeCheckDecl.cpp (hunk 4 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckDecl.cpp]: Oh, of course it is! Thanks.

**Comment by DougGregor** [comment]
And... had to rebase. Comments addressed, `SpecializedProtocolConformance` and `SILBoxType` converted. `ConcreteDeclRef` flattened.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Pull request merged**